Track widget creation locations.

jacob314 opened the pull request:
Run an optional kernel transform to rewrite all calls to create widgets to track the location the call was made from.

jacob314 commented:
@devoncarew @pq 
This is the Kernel Transformer portion of
https://github.com/flutter/flutter-intellij/issues/1597
There is a forthcoming change to IntelliJ and flutter_tools as well.
This CL is a little rough around the edges but I'd like some review feedback before I polish it too much in the wrong directions.

jacob314 commented:
I'd also appreciate advice on the most productive way to write tests for a kernel transformer. I could imagine an end to end test on the flutter side or adding more hooks to this code so it can apply the transform in a useful way without involving flutter code.

Hixie commented on frontend_server/lib/server.dart:
```diff
@@ -45,7 +47,14 @@ ArgParser _argParser = new ArgParser(allowTrailingOptions: true)
+      'When in batch mode, link platform kernel file into result kernel file.'
+      ' Intended use is to satisfy different loading strategies implemented'
+      ' by gen_snapshot(which needs platform embedded) vs'
+      ' Flutter engine(which does not)',
```
trivial nits: spaces before `(`, also, put the space at the end of the line rather than the start, also, line up the strings so their open `'`s are all vertically aligned.

Hixie commented on frontend_server/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,511 @@
+              initializer.arguments,
+              initializer.target.function,
+              new VariableGet(variable),
+              _locationClass);
```
is this how dartfmt formatted this?

jacob314 commented:
Yes. Formatted with dartfmt.

On Jan 9, 2018 8:31 PM, "Ian Hickson" <notifications@github.com> wrote:

> *@Hixie* commented on this pull request.
> ------------------------------
>
> In frontend_server/lib/track_widget_constructor_locations.dart
> <https://github.com/flutter/engine/pull/4529#discussion_r160587933>:
>
> > +            // not existing.
> +            handleConstructor(initializer.target);
> +          }
> +
> +          _maybeAddCreationLocationArgument(
> +              initializer.arguments,
> +              initializer.target.function,
> +              new VariableGet(variable),
> +              _locationClass);
> +        } else if (initializer is SuperInitializer &&
> +            _instanceOfWidget(initializer.target.enclosingClass)) {
> +          _maybeAddCreationLocationArgument(
> +              initializer.arguments,
> +              initializer.target.function,
> +              new VariableGet(variable),
> +              _locationClass);
>
> is this how dartfmt formatted this?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/engine/pull/4529#pullrequestreview-87730233>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABK4PGRaq3W9NRWqMzcbVnsJR4MOxyPiks5tJD09gaJpZM4RYLhP>
> .
>

devoncarew commented on frontend_server/lib/server.dart:
```diff
@@ -45,7 +47,14 @@ ArgParser _argParser = new ArgParser(allowTrailingOptions: true)
+      defaultsTo: true)
+  ..addFlag('track-widget-creation-locations',
+      help: 'Run a kernel transformer to track creation locations for widgets '
+      'When in batch mode, link platform kernel file into result kernel file.'
```
Do you want this `When in batch mode...` text? It seems to go with the proceeding cli flag.

jacob314 replied:
Doh. Removed.

devoncarew commented on frontend_server/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,511 @@
+library track_widget_constructor_locations;
```
copyright header?

jacob314 replied:
done

devoncarew commented on frontend_server/lib/server.dart:
```diff
@@ -45,7 +47,14 @@ ArgParser _argParser = new ArgParser(allowTrailingOptions: true)
           ' Flutter engine(which does not)',
-      defaultsTo: true);
+      defaultsTo: true)
+  ..addFlag('track-widget-creation-locations',
```
DBC: this flag name is long; perhaps `--track-widget-creation`? `--track-widgets`?

jacob314 replied:
--track-widget-creation is much better. Updated here and elsewhere.

aam commented on frontend_server/lib/server.dart:
```diff
@@ -227,6 +255,7 @@ class _FrontendCompiler implements CompilerInterface {
     return Uri.base.resolve(path);
   }
 
+
```
nit: here and below new blank lines seem unintended

jacob314 replied:
removed. cleaned up

aam commented on frontend_server/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,511 @@
+/// This transformer requires that all Widget constructors have already been
+/// transformed to have a named parameter with the name specified by
+/// `_locationParameterName`.
+class _WidgetCallSiteTransformer extends Transformer {
```
Like I said on the email chain I feel uneasy about putting kernel transforms into frontend_server directly. Ideally we keep frontend_server deal with incremental, staying-in-memory copy of frontend only, find different place for general flutter-specific kernel transforms.

aam replied:
I would expect somebody from @kmillikin to be better reviewers of the transformer code.

jacob314 replied:
@kmillikin can you please have someone on your team review the transformer code?
I've moved the kernel transformers code to a flutter_kernel_transformers package
which can be a central location for flutter kernel transformers. Added flutter to the name as otherwise the import from the frontent_server is surprising.

Hixie commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,516 @@
+library track_widget_constructor_locations;
```
needs license/copyright block like other dart files

jacob314 replied:
done. also added one to server.dart which was missing one.

Hixie commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,516 @@
+  /// of the parameters passed in so that tools can show parameter locations
+  /// without reparsing the source code.
+  ConstructorInvocation _constructLocation(Location location,
+      {String name, ListLiteral parameterLocations, bool showFile: true}) {
```
to make dartfmt do a better job of this, you can put a comma before the `}`

jacob314 replied:
done.

Hixie commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,516 @@
+          constructor.function, _creationLocationParameterName));
+      final VariableDeclaration variable = new VariableDeclaration(
+          _creationLocationParameterName,
+          type: _locationClass.thisType);
```
same here (comma before the `)`)

jacob314 replied:
done

Hixie commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,516 @@
+              initializer.arguments,
+              initializer.target.function,
+              new VariableGet(variable),
+              _locationClass);
```
ditto

Hixie replied:
there are others in this file. Basically any time dartfmt ends up doing weird things where you can't tell the nesting at a glance from the indentation, it's a good place to try adding commas until it does a better job.

jacob314 replied:
done

jacob314 replied:
done

Hixie commented:
@aam it strikes me that having these dart programs and libraries in the engine repo means that we're not pinning down their dependencies the way we do for other flutter tools... :-( I guess we should update the script that pins dependencies to also look in the engine repo...

aam commented:
@Hixie wrote

> @aam it strikes me that having these dart programs and libraries in the engine repo means that we're not pinning down their dependencies the way we do for other flutter tools... :-( I guess we should update the script that pins dependencies to also look in the engine repo...

Yes, good point. But should that script include dart projects from flutter/engine repo when it runs against flutter/flutter or should it pin engine repo projects dependencies separately from flutter/flutter? I got impression that flutter/engine and flutter/flutter are loosely coupled, run on different dart sdk even.

jacob314 commented:
@aam @Hixie It would be nice if these transformers could be in the flutter repo instead of the engine repo but I can't think of a clean way to support that.

Hixie commented:
> I got impression that flutter/engine and flutter/flutter are loosely coupled, run on different dart sdk even.

We run the compiler server with a different Dart SDK than the flutter_tools tool? Aren't they running on the same host next to each other?

aam commented:
> We run the compiler server with a different Dart SDK than the flutter_tools tool? Aren't they running on the same host next to each other?

Yes, they run on different sdks despite being next to each other.

Hixie commented:
Actually maybe the better solution then is to just make sure we only use packages from the dart SDK in these engine-side Dart apps (none from pub, where we might introduce skew across different tools). It looks like that's close to what we're doing, I only see one or two from pub right now.

stefantsov commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,567 @@
+  }
+
+  bool _instanceOfWidget(Class clazz) {
+    return _hierarchy.getClassAsInstanceOf(clazz, _widgetClass) != null;
```
The same result can be achieved by `_hierarchy.isSubtypeOf(clazz, _widgetClass)` which may be a bit more readable.

jacob314 replied:
done

stefantsov commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,567 @@
+    // argument to the factory constructor rather than the location
+    if (_currentFactory != null &&
+        _hierarchy.getClassAsInstanceOf(
+                constructedClass, _currentFactory.enclosingClass) !=
```
As pointed above, `isSubtypeOf` may be a bit more readable than `getClassAsInstanceOf`.

jacob314 replied:
done. that is better. I missed that method scanning through the api.

stefantsov commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,567 @@
+    if (clazz == null) {
+      return false;
+    }
+    return hierarchy.getClassAsInstanceOf(clazz, _widgetClass) != null;
```
As pointed above, `hierarchy.isSubtypeOf(clazz, _widgetClass)` may be a bit more readable.

jacob314 replied:
done

jacob314 replied:
I switched these methods but the code fails because the methods are only on ClosedWorldClassHierarchy not on IncrementalClassHierarchy which I believe I need to use.
I'm happy to refactor my code if the helper method is added to the base ClassHierarchy class.

stefantsov replied:
Yes, sorry for that confusion.  ClassHierarchy and related classes are being changed currently.  In short, IncrementalClassHierarchy is going to be deprecated (you may already see the related changes in the current SDK checkout), and I strongly recommend to avoid using it.  If that's possible to use the latest changes from the Dart SDK's master branch, I would suggest the following for now:

- Don't update the class hierarchy, create the new one when the classes are updated instead.
- When needed, create a ClassHierarchy using [the recently added factory](https://github.com/dart-lang/sdk/blob/master/pkg/kernel/lib/class_hierarchy.dart#L20).
- When [CL 35700](https://dart-review.googlesource.com/c/sdk/+/35700) is landed (I hope it's going to be today), ClassHierarchy will have isSubtypeOf, and you will be able to use it here and in other analogous places of the transformer.

We do plan to add a component-based class hierarchy, so that it won't be necessary to recreate it in order to update it.

stefantsov commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,567 @@
+    final Field locationField = new Field(
+      new Name(
+        _locationFieldName,
+        _hasCreationLocationClass.enclosingLibrary,
```
Unfortunately, it's hard to guess from the comment right above if the desired behavior can be achieved by using the enclosing library of `_hasCreationLocationClass` instead of the enclosing library of `clazz`.

jacob314 replied:
Comment was out of date. It needs to be updated to reflect that I really do want the library of the  _hasCreationLocationClass which defines this private getter _location.

stefantsov commented on flutter_kernel_transformers/lib/track_widget_constructor_locations.dart:
```diff
@@ -0,0 +1,567 @@
+
+    // Handle the widget class and classes that implement but do not extend the
+    // widget class.
+    if (_instanceOfWidget(clazz.superclass)) {
```
I think classes that extend the widget class will pass this check as well.

jacob314 replied:
good catch. this check got reversed accidentally when refactoring. The idea is if the superclass does not implement Widget then this class by definition implements Widget or is the widget class.

jacob314 commented:
PTAL. I've uploaded a new version.

jacob314 commented:
Friendly ping. I'd really like to get this change landed soon so Widget
Inspector users can take advantage of this really useful information.

On Tue, Jan 16, 2018 at 9:20 PM, Jacob Richman <notifications@github.com>
wrote:

> PTAL. I've uploaded a new version.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/engine/pull/4529#issuecomment-358200322>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABK4PN1fE_VSAiFXigT3c58osiTujOi3ks5tLYMtgaJpZM4RYLhP>
> .
>

Hixie commented on flutter_kernel_transformers/pubspec.lock:
```diff
@@ -0,0 +1,335 @@
+# Generated by pub
```
this is a library, right? it shouldn't need a pubspec.lock

jacob314 replied:
it is a package same as frontend_server. That was the design @aam aam recommended. I don't have a strong preference once way or the other.

jacob314 replied:
removed pubspec.lock file. it shouldn't be in the CL.

Hixie replied:
I believe the difference between frontend_server and flutter_kernel_transformers is that the former is an app (it's run directly) whereas the latter is a library (it's imported by the other). Is that an incorrect interpretation?

Hixie commented on frontend_server/.packages:
```diff
@@ -1,4 +1,4 @@
-# Generated by pub on 2017-08-22 07:05:11.485224.
```
this file is gone now

jacob314 replied:
removed from CL.

Hixie commented on travis/analyze.sh:
```diff
@@ -12,6 +12,20 @@ if [ -n "$RESULTS" ]; then
+  --packages=flutter/flutter_kernel_transformers/.packages                                 \
+  --options flutter/analysis_options.yaml                                      \
+  flutter/flutter_kernel_transformers                                                      \
+  2>&1                                                                         \
```
indentation on the right side is off

jacob314 replied:
done

Hixie commented:
Since this is dealing with code I'm not familiar with, I will defer to one of the Dart folks for the final stamp.

chinmaygarde commented:
Trying to clear the PR queue. What is the status of this patch?

jacob314 commented:
I will be updating this CL to account for the changes in https://dart-review.googlesource.com/c/sdk/+/35700 and landing this week.

jacob314 commented:
Can I have an LGTM? All tests now pass and I've updated to use take advantage of the new ClassHierarchy changes.

aam commented on frontend_server/lib/server.dart:
```diff
@@ -227,6 +253,7 @@ class _FrontendCompiler implements CompilerInterface {
     return Uri.base.resolve(uriPath);
   }
 
+
```
accidentally added newline?

aam commented on frontend_server/lib/server.dart:
```diff
@@ -237,6 +264,7 @@ class _FrontendCompiler implements CompilerInterface {
   }
 }
 
+
```
accidentally added newline?

jacob314 replied:
fixed

jacob314 commented:
Added missing entry for flutter_kernel_transformers in frontend_server/BUILD.gn
Switched back to the deprecated ClassHierarchy constructor due to a bug
https://github.com/dart-lang/sdk/issues/32079
Once that bug has been fixed I'll switch to the non-deprecated constructor.
